pkg/operator/encryption/prune_controller.go

```diff
@@ -0,0 +1,226 @@
+	encryptedGRs map[schema.GroupResource]bool,
+) *pruneController {
+	c := &pruneController{
+		operatorClient: operatorClient,
+
+		queue:         workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "EncryptionPruneController"),
+		eventRecorder: eventRecorder.WithComponentSuffix("encryption-prune-controller"), // TODO unused
+
+		encryptedGRs:    encryptedGRs,
+		targetNamespace: targetNamespace,
+
+		encryptionSecretSelector: encryptionSecretSelector,
+		podClient:                podClient,
+		secretClient:             secretClient,
+	}
+
+	c.preRunCachesSynced = setUpAllEncryptionInformers(operatorClient, targetNamespace, kubeInformersForNamespaces, c.eventHandler())
+
+	return c
+}
+
+func (c *pruneController) sync() error {
+	if ready, err := shouldRunEncryptionController(c.operatorClient); err != nil || !ready {
+		return err // we will get re-kicked when the operator status updates
+	}
+
+	// TODO do we want to use this to control the number we keep around?
+	// operatorSpec.SucceededRevisionLimit
+
+	configError := c.deleteOldMigratedSecrets()
+
+	// update failing condition
+	cond := operatorv1.OperatorCondition{
+		Type:   "EncryptionPruneControllerDegraded",
+		Status: operatorv1.ConditionFalse,
+	}
+	if configError != nil {
+		cond.Status = operatorv1.ConditionTrue
+		cond.Reason = "Error"
+		cond.Message = configError.Error()
+	}
+	if _, _, updateError := operatorv1helpers.UpdateStaticPodStatus(c.operatorClient, operatorv1helpers.UpdateStaticPodConditionFn(cond)); updateError != nil {
+		return updateError
+	}
+
+	return configError
+}
+
+func (c *pruneController) deleteOldMigratedSecrets() error {
+	_, encryptionState, _, err := getEncryptionConfigAndState(c.podClient, c.secretClient, c.targetNamespace, c.encryptionSecretSelector, c.encryptedGRs)
+	if err != nil {
+		return err
+	}
+	// do not try to delete anything during a transition state
+	if len(encryptionState) == 0 {
+		c.queue.AddAfter(pruneWorkKey, 2*time.Minute)
+		return nil
+	}
+
+	usedSecrets := make([]*corev1.Secret, 0, len(encryptionState))
+	for _, grKeys := range encryptionState {
+		usedSecrets = append(usedSecrets, grKeys.readSecrets...)
+	}
+
+	allkeys, err := c.secretClient.Secrets(operatorclient.GlobalMachineSpecifiedConfigNamespace).List(c.encryptionSecretSelector)
+	if err != nil {
+		return err
+	}
+
+	// sort by keyID
+	encryptionSecrets := make([]*corev1.Secret, 0, len(allkeys.Items))
+	for _, s := range allkeys.Items {
+		encryptionSecrets = append(encryptionSecrets, s.DeepCopy()) // don't use &s because it constant through-out the loop
+	}
+	sort.Slice(encryptionSecrets, func(i, j int) bool {
+		// it is fine to ignore the validKeyID bool here because we filtered out invalid secrets in the loop above
+		iKeyID, _ := secretToKeyID(encryptionSecrets[i])
+		jKeyID, _ := secretToKeyID(encryptionSecrets[j])
+		return iKeyID > jKeyID
+	})
+
+	var deleteErrs []error
+	skippedKeys := 0
+	for _, s := range encryptionSecrets {
```

Comment:
do we want to collect and dump (to the log file) the number of keys that were deleted?

Reply:
done